File: benchmarks/python/tree_traversal.py
#!/usr/bin/env python3

# Tree Traversal benchmark in Python

class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

def build_tree(depth, val=0):
    if depth <= 0:
        return None
    node = TreeNode(val)
    node.left = build_tree(depth - 1, val * 2 + 1)
    node.right = build_tree(depth - 1, val * 2 + 2)
    return node

def inorder_traversal(node, result):
    if node:
        inorder_traversal(node.left, result)
        result.append(node.val)
        inorder_traversal(node.right, result)

def preorder_traversal(node, result):
    if node:
        result.append(node.val)
        preorder_traversal(node.left, result)
        preorder_traversal(node.right, result)

def postorder_traversal(node, result):
    if node:
        postorder_traversal(node.left, result)
        postorder_traversal(node.right, result)
        result.append(node.val)

# Build tree of depth 15
root = build_tree(15)

# Traverse
inorder_result = []
inorder_traversal(root, inorder_result)

preorder_result = []
preorder_traversal(root, preorder_result)

postorder_result = []
postorder_traversal(root, postorder_result)

print(f"Tree nodes: {len(inorder_result)}, Sum: {sum(inorder_result)}")

File: benchmarks/javascript/tree_traversal.js
// Tree Traversal benchmark in JavaScript

class TreeNode {
    constructor(val = 0) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

function buildTree(depth, val = 0) {
    if (depth <= 0) {
        return null;
    }
    const node = new TreeNode(val);
    node.left = buildTree(depth - 1, val * 2 + 1);
    node.right = buildTree(depth - 1, val * 2 + 2);
    return node;
}

function inorderTraversal(node, result) {
    if (node) {
        inorderTraversal(node.left, result);
        result.push(node.val);
        inorderTraversal(node.right, result);
    }
}

function preorderTraversal(node, result) {
    if (node) {
        result.push(node.val);
        preorderTraversal(node.left, result);
        preorderTraversal(node.right, result);
    }
}

function postorderTraversal(node, result) {
    if (node) {
        postorderTraversal(node.left, result);
        postorderTraversal(node.right, result);
        result.push(node.val);
    }
}

// Build tree of depth 15
const root = buildTree(15);

// Traverse
const inorderResult = [];
inorderTraversal(root, inorderResult);

const preorderResult = [];
preorderTraversal(root, preorderResult);

const postorderResult = [];
postorderTraversal(root, postorderResult);

const sum = inorderResult.reduce((a, b) => a + b, 0);
console.log(`Tree nodes: ${inorderResult.length}, Sum: ${sum}`);
